typing/typedecl_separability.ml

```diff
@@ -0,0 +1,674 @@
+         an accumulator variable [acc], setting existential variables
+         to Ind as we go. *)
+      let param_instance = Ctype.repr param_instance in
+      let get context var =
+        try TVarMap.find var context with Not_found -> Ind in
+      let set_ind context var =
+        TVarMap.add var Ind context in
+      let is_ind context var = match get context var with
+        | Ind -> true
+        | Sep | Deepsep -> false in
+      match param_instance.desc with
+      | Tvar text ->
+          let var = {text; id = param_instance.Types.id} in
+          (get context var) :: acc, (set_ind context var)
+      | _ ->
+          let instance_exis = free_variables param_instance in
+          if List.for_all (is_ind context) instance_exis then
+            Ind :: acc, context
+          else
+            Deepsep :: acc, List.fold_left set_ind context instance_exis
+    in
+    let mode_signature, context =
+      (* fold_right here is necessary to get the mode
+         consed to the accumulator in the right order *)
+      List.fold_right handle_equation parameters ([], context) in
+    (* After all variable determined by the parameters have been set to Ind
+       by [handle_equation], all variables remaining in the context are
+       purely existential and should not require a stronger mode than Ind. *)
+    let check_existential evar mode =
+      if rank mode > rank Ind then
+        raise (Error (decl_loc, Non_separable_evar evar.text))
+    in
+    TVarMap.iter check_existential context;
+    mode_signature
+
+(** [check_def env def] returns the signature required
+    for the type definition [def] in the typing environment [env].
+
+    The exception [Not_separable] is raised if we discover that
+    no such signature exists -- the definition will always be invalid. This
+    only happens when the definition is marked to be unboxed. *)
+let check_def
+  : Env.t -> type_definition -> Sep.signature
+  = fun env def ->
+  let boxed = not def.type_unboxed.unboxed in
+  match structure def with
+  | Abstract ->
+      assert boxed;
+      msig_of_external_type def
+  | Synonym type_expr ->
+      check_type env type_expr Sep
+      |> msig_of_context ~decl_loc:def.type_loc ~parameters:def.type_params
+  | Open | Algebraic (Zero | Several | One (Zero | Several)) ->
+      assert boxed;
+      best_msig def
+  | Algebraic (One (One constructor)) ->
+    if boxed then best_msig def
+    else
+      check_type env constructor.argument_type Sep
+      |> msig_of_context ~decl_loc:def.type_loc
+           ~parameters:constructor.result_type_parameter_instances
+
+let compute_decl env decl =
+  if not Config.flat_float_array then best_msig decl
+  else check_def env decl
+
+
+(** Separability as a generic property *)
+type prop = Types.Separability.signature
+
+let property : (prop, unit) Typedecl_properties.property =
+  let open Typedecl_properties in
+  let eq ts1 ts2 = try List.for_all2 Sep.eq ts1 ts2 with _ -> false in
```

Comment:
```suggestion
  let eq ts1 ts2 = try List.for_all2 Sep.eq ts1 ts2 with Invalid_argument _ -> false in
```

`try ... with _ ` is bad, see the whole discussion about asynchronous exceptions.